metaelf/metaelf.c

```diff
@@ -0,0 +1,264 @@
+
+
+static struct {
+	const char *name;
+	union {
+		void *memptr;
+		unsigned char *ident;
+		Elf32_Ehdr *hdr32;
+		Elf64_Ehdr *hdr64;
+	} elf;
+	int quiet;
+	enum {
+		mode_checkCRC,
+		mode_writeCRC,
+	} mode;
+} common;
+
+
+extern uint32_t crc32_calc(const uint8_t *buf, uint32_t len, uint32_t base);
+
+
+static uint16_t uint16(uint16_t val)
+{
+	return (common.elf.ident[EI_DATA] == ENDIANNESS) ? val : __builtin_bswap16(val);
+}
+
+
+static uint32_t uint32(uint32_t val)
+{
+	return (common.elf.ident[EI_DATA] == ENDIANNESS) ? val : __builtin_bswap32(val);
+}
+
+
+static uint64_t uint64(uint64_t val)
+{
+	return (common.elf.ident[EI_DATA] == ENDIANNESS) ? val : __builtin_bswap64(val);
+}
+
+
+static size_t elf32_size(Elf32_Ehdr *e)
+{
+	return uint32(e->e_shoff) + uint16(e->e_shentsize) * uint16(e->e_shnum);
+}
+
+
+static size_t elf64_size(Elf64_Ehdr *e)
+{
+	return uint64(e->e_shoff) + uint16(e->e_shentsize) * uint16(e->e_shnum);
+}
+
+
+/* Returns ELF file size based on its metadata */
+static ssize_t elf_size(void)
+{
+	const int elf_class = common.elf.ident[EI_CLASS];
+
+	if (elf_class == ELFCLASS32) {
+		return elf32_size(common.elf.hdr32);
+	}
+	else if (elf_class == ELFCLASS64) {
+		return elf64_size(common.elf.hdr64);
+	}
+
+	return -1;
+}
+
+
+/* Calculate ELF file checksum in host's byte order */
+static uint32_t elf_calc_crc32(uint8_t *ptr, size_t sz, size_t ofs)
+{
+	uint8_t zero[4] = { 0 };
+	uint32_t crc = (uint32_t)-1;
+	crc = crc32_calc(ptr, ofs, crc);
+	crc = crc32_calc(zero, 4, crc);
+	crc = crc32_calc(ptr + ofs + 4, sz - (ofs + 4), crc);
+
+	return ~crc;
+}
+
+
+static int parse_args(int argc, char **argv)
+{
+	int opt;
+
+	while ((opt = getopt(argc, argv, "qwch")) != -1) {
+		switch (opt) {
+			case 'q':
+				common.quiet = 1;
+				break;
+
+			case 'w':
+				common.mode = mode_writeCRC;
+				break;
+
+			case 'c':
+				common.mode = mode_checkCRC;
+				break;
+
+			case 'h': /* fall-through */
+			default:
+				printf(
+					"Usage: %s [OPTIONS] <file.elf>\n"
+					"Options:\n"
+					"  -h   Prints this help\n"
+					"  -c   Check ELF CRC32 with embedded checksum (default)\n"
+					"  -w   Embed CRC32 into ELF file header\n"
+					"  -q   Silent mode\n",
+					basename(argv[0]));
+				return -1;
+		}
+	}
+
+	if (argc - optind != 1) {
+		log_error("No input file");
+		return -1;
+	}
+
+	common.name = argv[optind];
+
+	return 0;
+}
+
+
+int main(int argc, char **argv)
+{
+	struct stat st = { 0 };
+	uint32_t crcOut, crcIn = 0;
+	int fd, ret = EXIT_FAILURE;
+
+	common.elf.memptr = MAP_FAILED;
+
+	if (parse_args(argc, argv) < 0) {
+		return EXIT_FAILURE;
+	}
+
+	fd = open(common.name, O_RDWR);
+	if (fd < 0) {
+		log_error("Unable to open file %s", common.name);
+		return EXIT_FAILURE;
+	}
+
+	do {
+		if (fstat(fd, &st) == -1) {
+			break;
+		}
+
+		if (st.st_size == 0) {
+			log_error("File has a zero size");
+			break;
+		}
+
+		common.elf.memptr = mmap(0, st.st_size, PROT_READ | PROT_WRITE, MAP_SHARED, fd, 0);
```

Comment:
This tool is `to be used on host` only during system image build process (when adding programs to the syspage) on targets which does not provide filesystem capable to provide integrity checks.

I agree that some mmaps' arguments should be aligned to `PAGE_SIZE`, but `length` (second argument) describes the `minimum` amount of bytes the mapping will span and in case if `fd` is used it also affects the file size on write. Of course since `mmap()` maps pages into memory, the `length` may be rounded up to hit a page boundary, but on disk write we write exactly `length` bytes. The requirement for arguments to provide necessary alignment is for `address` (NULL here) and `offset` (0 here).

Btw, I've added `byteswap helper` to be able to build this tool on *BSD's and Darwin.

Reply:
It's ok, I only noticed some potential compatibility issues if you'd also like to run this on target, otherwise it's totally ok